utils/fixed_decimal/src/decimal.rs

```diff
@@ -697,6 +718,495 @@ impl FromStr for FixedDecimal {
+    fn round_trailing_digits(&mut self, n: u16, mode: RoundingMode) -> Result<(), Error> {
+        /// For fixing up invariants after truncation
+        ///
+        /// Basically will ensure that we don't end with any zeroes
+        fn fixup_invariants(dec: &mut FixedDecimal) {
+            let first_nonzero = dec.digits.iter().rposition(|d| *d != 0).unwrap_or(0);
+            dec.digits.truncate(first_nonzero + 1);
+            if dec.digits.is_empty() {
+                dec.magnitude = 0;
+                dec.upper_magnitude = 0;
+            }
+        }
+
+        debug_assert!(
+            self.digits.len() >= n as usize,
+            "Attempted to round off {} digits of number that has only {}",
+            n,
+            self.digits.len()
+        );
+        if n == 0 {
+            // no point attempting to round off any digits
+            #[cfg(debug_assertions)]
+            self.check_invariants();
+            return Ok(());
+        }
+
+        let cutoff = self.digits.len() - n as usize;
+
+        match mode {
+            RoundingMode::Unnecessary => {
+                // If we got to this point then rounding was not unnecessary
+                return Err(Error::Limit);
+            }
+            RoundingMode::Truncate => {
+                self.digits.truncate(cutoff);
+                fixup_invariants(self);
+                #[cfg(debug_assertions)]
+                self.check_invariants();
+                return Ok(());
+            }
+            // continue to rest of routine
+            RoundingMode::HalfExpand => (),
+        }
+
+        // Do we need to round our significant digits?
+        // TODO(#1177): This heuristic is insufficient for most rounding modes.
+        let round = self.digits[cutoff] >= 5;
+
+        self.digits.truncate(cutoff);
+
+        if round {
+            // how much to truncate by after rounding
+            let mut round_truncate = cutoff;
+            for digit in self.digits[..cutoff].iter_mut().rev() {
+                if *digit == 9 {
+                    // Truncate this digit, the next digit can be rounded
+                    round_truncate -= 1;
+                } else {
+                    // We need to update this digit, then we're done
+                    *digit += 1;
+                    self.digits.truncate(round_truncate);
+                    #[cfg(debug_assertions)]
+                    self.check_invariants();
+                    return Ok(());
+                }
+            }
+
+            // If we reached this point then the last digit was 9 and we need to insert
+            // another digit at the beginning
+
+            self.digits.clear();
+            self.digits.push(1);
+            debug_assert!(self.upper_magnitude >= 0);
+            self.magnitude += 1;
+            if self.upper_magnitude < self.magnitude {
+                self.upper_magnitude = self.magnitude;
+            }
+        }
+        fixup_invariants(self);
```

Comment:
Nit: I think you can safely put the `fixup_invariants` into the `else` of the `if round` that precedes this